### PR TITLE
AWS: Support checksum validation with S3 eTags

### PR DESCRIPTION
**Pull request by SinghAsDev**
This PR adds capability to perform checksum validations on data uploads using S3 eTags. This should help in capturing hardware issues like corrupt disks that may affect the staged files generated while uploading data to S3.

### CODE REVIEW

**Comment by SinghAsDev** [comment]
@jackye1995 mind reviewing this when you get a chance.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 1]
```diff
@@ -209,6 +209,12 @@
    */
   public static final String CLIENT_ASSUME_ROLE_REGION = "client.assume-role.region";
 
+  /**
+   * Enables eTag checks for S3 PUT and MULTIPART upload requests.
+   */
+  public static final String CLIENT_ENABLE_ETAG_CHECK = "client.enable.etag-check";
```
Comment (kbendick) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:215]: Would it make more sense to use `client.etag.enable-check` or something with `etag` as the second part?

Some of the other ones seem to follow a pattern of "client".<resource>.<configuration>, which I think makes more sense.

**Comment by wizardxz** [comment]
LGTM

**Review by wizardxz — APPROVED** [review]

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 1 above)
Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:215]: since this is only for s3, I think it makes more sense to use `s3` prefix. Also the convention of naming for this type of config is `xxx.enabled` (see example like `SPARK_WRITE_PARTITIONED_FANOUT_ENABLED`)

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 2]
```diff
@@ -132,6 +145,11 @@ public void write(int b) throws IOException {
     }
 
     stream.write(b);
+    if (isEtagCheckEnabled) {
+      byte byteValue = ((Integer) b).byteValue();
+      currentPartMessageDigest.update(byteValue);
+      completeMessageDigest.update(byteValue);
+    }
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:152]: nit: newline after if block

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 3]
```diff
@@ -153,6 +171,10 @@ public void write(byte[] b, int off, int len) throws IOException {
       int writeSize = multiPartSize - (int) stream.getCount();
 
       stream.write(b, relativeOffset, writeSize);
+      if (isEtagCheckEnabled) {
+        currentPartMessageDigest.update(b, relativeOffset, writeSize);
+        completeMessageDigest.update(b, relativeOffset, writeSize);
+      }
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:177]: nit: newline after if block

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 4]
```diff
@@ -74,13 +77,17 @@
...
   private CountingOutputStream stream;
-  private final List<File> stagingFiles = Lists.newArrayList();
+  private final List<FileWithEtag> stagingFilesWithETags = Lists.newArrayList();
   private final File stagingDirectory;
   private File currentStagingFile;
   private String multipartUploadId;
   private final Map<File, CompletableFuture<CompletedPart>> multiPartMap = Maps.newHashMap();
   private final int multiPartSize;
   private final int multiPartThresholdSize;
+  private static final String digestAlgorithm = "MD5";
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:87]: nit: static variable should be above all other non-static ones

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 5]
```diff
@@ -74,13 +77,17 @@
   private final AwsProperties awsProperties;
 
   private CountingOutputStream stream;
-  private final List<File> stagingFiles = Lists.newArrayList();
+  private final List<FileWithEtag> stagingFilesWithETags = Lists.newArrayList();
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:80]: I think we can avoid changing this variable name because the class name is already self-explanatory

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 6]
```diff
@@ -178,7 +204,12 @@ private void newStream() throws IOException {
     createStagingDirectoryIfNotExists();
     currentStagingFile = File.createTempFile("s3fileio-", ".tmp", stagingDirectory);
     currentStagingFile.deleteOnExit();
-    stagingFiles.add(currentStagingFile);
+    try {
+      currentPartMessageDigest = isEtagCheckEnabled ? MessageDigest.getInstance(digestAlgorithm) : null;
+    } catch (NoSuchAlgorithmException e) {
+      throw new IOException(e);
+    }
+    stagingFilesWithETags.add(new FileWithEtag(currentStagingFile, currentPartMessageDigest));
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:212]: nit: newline after try block

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 7]
```diff
@@ -216,43 +247,43 @@ private void uploadParts() {
...
 
           S3RequestUtil.configureEncryption(awsProperties, requestBuilder);
 
           UploadPartRequest uploadRequest = requestBuilder.build();
 
           CompletableFuture<CompletedPart> future = CompletableFuture.supplyAsync(
               () -> {
-                UploadPartResponse response = s3.uploadPart(uploadRequest, RequestBody.fromFile(f));
+                UploadPartResponse response = s3.uploadPart(uploadRequest, RequestBody.fromFile(f.file()));
+                checkEtag(f.eTag(), response.eTag());
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:270]: S3 has built in checksum verification using request header `Content-MD5`. I think we don't need to check the response at client side.

See https://aws.amazon.com/premiumsupport/knowledge-center/data-integrity-s3/ for more details

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 1 above)
Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:215]: The name "etag check" feels a bit not informative to me, people without context has no idea what this is doing. Maybe something like `s3.integrity-check.enabled` or `s3.checksum.enabled` sounds better?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 7 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:270]: Yea, I was also initially doing that. However, I could not find a way to add a reliable test. Tests would succeed with wrong md5 checksums added to request. Likely due to s3 mock. With that and the fact that current approach allows to add a better error message, I would propose doing explicit checks here. However, if you have a strong preference on this, I can modify. Let me know.

Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:270]: @jackye1995 thoughts?

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 1 above)
Comment (kbendick) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:215]: I'm a big fan of `s3` as the prefix, and then `s3.checksum.enabled` would be my preferred choice.

This really explains what the intent is, vs etag-check more explains how it works and isn't as informative for users who don't know the underlying details of S3.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 7 above)
Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:270]: The integration tests in `/aws/src/integration` are added for such a purpose, I think we can leverage that.

Unfortunately it is not run as a part of CI yet because it requires connection to an AWS account, we are working on getting one for Iceberg. If you do not have any AWS infrastructure that could be used to run it, I can run it for you after you publish you tests in this PR.

Comment (jackye1995) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:270]: > With that and the fact that current approach allows to add a better error message, I would propose doing explicit checks here.

Consulted with the S3 team related to this, definitely doing it on server side is preferred as they have dedicated hardware for this operation, comparing to a client running checksum in any environment

Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:270]: Thanks @jackye1995 . Let me update the diff in that case. Will try to do so this week.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java (hunk 1 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java]: SGTM, updated to `s3.checksum.enabled`

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 7 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java]: @jackye1995 I think this is good now. I verified that wrong md5 is causing protocol exceptions from s3 server side.

**Comment by rdblue** [comment]
@danielcweeks, can you take a look at this?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 8]
```diff
@@ -178,7 +208,13 @@ private void newStream() throws IOException {
     createStagingDirectoryIfNotExists();
     currentStagingFile = File.createTempFile("s3fileio-", ".tmp", stagingDirectory);
     currentStagingFile.deleteOnExit();
-    stagingFiles.add(currentStagingFile);
+    try {
+      currentPartMessageDigest = isEtagCheckEnabled ? MessageDigest.getInstance(digestAlgorithm) : null;
+    } catch (NoSuchAlgorithmException e) {
+      throw new IOException(e);
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:214]: Why does this throw `IOException` with no context? I think there should be context at a minimum. I also suspect that this is throw `IOException` because that's an allowed checked exception?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 9]
```diff
@@ -360,4 +421,22 @@ protected void finalize() throws Throwable {
       LOG.warn("Unclosed output stream created by:\n\t{}", trace);
     }
   }
+
+  private static class FileWithEtag {
+    private final File file;
+    private final MessageDigest eTag;
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:427]: Why keep the `MessageDigest` around when the only use of it is to immediately call `digest()` on line 269? Seems like this should probably track the digest bytes with the file instead.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 10]
```diff
@@ -218,41 +254,49 @@ private void uploadParts() {
...
-        .filter(Predicates.not(multiPartMap::containsKey))
+        .filter(Predicates.not(f -> multiPartMap.containsKey(f.file())))
         .forEach(f -> {
           UploadPartRequest.Builder requestBuilder = UploadPartRequest.builder()
               .bucket(location.bucket())
               .key(location.key())
               .uploadId(multipartUploadId)
               .partNumber(stagingFiles.indexOf(f) + 1)
-              .contentLength(f.length());
+              .contentLength(f.file().length());
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:266]: Within this block, there would be a lot fewer changes if you left `f` unchanged: `.foreach(fileAndDigest -> { File f = fileAndDigest.file(); ... })`.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 11]
```diff
@@ -311,16 +356,32 @@ private void completeUploads() {
...
     } else {
       uploadParts();
       completeMultiPartUpload();
     }
   }
 
+  private static void checkProtocolException(UncheckedIOException uncheckedIOException, String requestString) {
+    if (uncheckedIOException.getCause() instanceof ProtocolException) {
+      LOG.error("S3 Request Failure: {}", requestString);
+      throw uncheckedIOException;
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:381]: This only throws protocol exceptions? What is being suppressed by this? Is there a guarantee that this is safe?

**Review by rdblue — CHANGES_REQUESTED** [review]
This looks close to being ready, but the error handling in `checkProtocolException` is concerning to me. It seems to swallow a broad range of exceptions.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 11 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:381]: Oh this is a mistake. Thanks for catching this @rdblue . Will fix it shortly

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 9 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:427]: I think we can do that, it will add the logic a tad bit more complicated as we close stream at multiple places. Will update it.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 12]
```diff
@@ -74,13 +78,16 @@
...
+  private final List<FileAndDigest> stagingFiles = Lists.newArrayList();
   private final File stagingDirectory;
   private File currentStagingFile;
   private String multipartUploadId;
   private final Map<File, CompletableFuture<CompletedPart>> multiPartMap = Maps.newHashMap();
   private final int multiPartSize;
   private final int multiPartThresholdSize;
+  private final boolean isEtagCheckEnabled;
+  private final MessageDigest completeMessageDigest;
+  private MessageDigest currentPartMessageDigest;
```
Comment (danielcweeks) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:90]: Rather than computing all of the digests ourselves, it might be cleaner to wrap the streams in `java.security.DigestOutputStream` and delegate to that in order to calculate the etag.  I'm not sure if that approach can be used for the complete message (since the parts are uploaded independently), but I also question whether we really need to validate the full checksum if we get all of the parts verified.

Comment (danielcweeks) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:90]: If we do want to calculate the full stream, it looks like it may be possible to wrap it twice: once for the individual part streams where we reset the digest every time, and a second wrapped stream where we reuse the digest across each file.

Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:90]: Thanks @danielcweeks for the review. We don't do full checksum if we are doing multi-part upload, we only do full checksum if it is a single put request. If I am understanding correctly wrapping the stream with `DigestOutputStream` will make us handle counts explicitly, which will beat the purpose of this suggestion to simplify the changes. Please correct me if I am missing something here.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 8 above)
Comment (rdblue) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java]: We don't want to change the signature, but we also don't want to throw some exception just because it was convenient. The best thing to do is to wrap in RuntimeException, which avoids throwing a misleading exception class.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 13]
```diff
@@ -218,41 +241,50 @@ private void uploadParts() {
...
             }
 
             if (thrown != null) {
               LOG.error("Failed to upload part: {}", uploadRequest, thrown);
               abortUpload();
             }
           });
-
-          multiPartMap.put(f, future);
+          multiPartMap.put(file, future);
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:287]: Can you revert the unnecessary whitespace change here?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 14]
```diff
@@ -360,4 +416,28 @@ protected void finalize() throws Throwable {
...
+
+    FileAndDigest(File file) {
+      this.file = file;
+    }
+
+    File file() {
+      return file;
+    }
+
+    void setDigest(byte[] digest) {
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:433]: Does this need to be mutable? Why not just pass in `digest` when creating? I thought that was possible before.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 15]
```diff
@@ -311,16 +344,39 @@ private void completeUploads() {
...
+      if (stagingFiles.size() > 0) {
+        stagingFiles.get(stagingFiles.size() - 1).setDigest(currentPartMessageDigest.digest());
+      }
+    }
+    stream.close();
+  }
+
+  private static void logS3RequestAndThrow(UncheckedIOException uncheckedIOException, String requestString) {
+    LOG.error("S3 Request Failure: {}", requestString);
+    throw uncheckedIOException;
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:377]: I don't think this method is needed. If the exception is going to be thrown, then it is the responsibility of whatever catches it to log it or handle it. This code doesn't know whether it can simply be handled, which is why it re-throws. So it should also not decide that this needs to be logged.

Patterns like this usually lead to misleading errors or red herrings in logs.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 16]
```diff
@@ -74,13 +79,16 @@
...
+  private final List<FileAndDigest> stagingFiles = Lists.newArrayList();
   private final File stagingDirectory;
   private File currentStagingFile;
   private String multipartUploadId;
   private final Map<File, CompletableFuture<CompletedPart>> multiPartMap = Maps.newHashMap();
   private final int multiPartSize;
   private final int multiPartThresholdSize;
+  private final boolean isEtagCheckEnabled;
+  private final MessageDigest completeMessageDigest;
+  private final MessageDigest currentPartMessageDigest;
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:91]: It looks like the purpose of having both digests is to handle the cases where either multi-part or PUT is used. Right?

Do we need both, or can we reuse the first part message digest?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 17]
```diff
@@ -360,4 +416,28 @@ protected void finalize() throws Throwable {
       LOG.warn("Unclosed output stream created by:\n\t{}", trace);
     }
   }
+
+  private static class FileAndDigest {
```
Comment (danielcweeks) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:420]: Could we make this extend File or make it more generic (FileReference) because we use this for both when we have a digest and when we don't.

This could simplify the etags checks on upload to just be:

```
If (file.hasDigest()) {
  requestBuilder.contentMD5(...)
}
```

It just feels like we're using the etag flag to drive a lot of logic in different places.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 18]
```diff
@@ -172,15 +187,23 @@ public void write(byte[] b, int off, int len) throws IOException {
...
     createStagingDirectoryIfNotExists();
     currentStagingFile = File.createTempFile("s3fileio-", ".tmp", stagingDirectory);
     currentStagingFile.deleteOnExit();
-    stagingFiles.add(currentStagingFile);
 
-    stream = new CountingOutputStream(new BufferedOutputStream(new FileOutputStream(currentStagingFile)));
+    stagingFiles.add(new FileAndDigest(currentStagingFile));
+
+    if (isEtagCheckEnabled) {
+      currentPartMessageDigest.reset();
```
Comment (danielcweeks) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:200]: This overlaps with Ryan's comment below, but if we just create a message digest and associate it here, we don't need to update it in the close method below.

Something like:

```java

stagingFiles.add(new FileAndDigest(currentStagingFile, newMessageDigest));
```

It means that we have to create a new digest object per file but also means they're standalone and don't require as much coordination.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 19]
```diff
@@ -373,4 +384,12 @@ public String dynamoDbTableName() {
   public void setDynamoDbTableName(String name) {
     this.dynamoDbTableName = name;
   }
+
+  public boolean isS3ChecksumEnabled() {
+    return this.isS3ChecksumEnabled;
+  }
+
+  public void setS3ChecksumEnabled(boolean eTagCheckEnabled) {
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:392]: nit: `eTagCheckEnabled` was not updated from the old name, maybe just `enabled`?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 20]
```diff
@@ -74,13 +79,16 @@
...
   private CountingOutputStream stream;
-  private final List<File> stagingFiles = Lists.newArrayList();
+  private final List<FileAndDigest> stagingFiles = Lists.newArrayList();
   private final File stagingDirectory;
   private File currentStagingFile;
   private String multipartUploadId;
   private final Map<File, CompletableFuture<CompletedPart>> multiPartMap = Maps.newHashMap();
   private final int multiPartSize;
   private final int multiPartThresholdSize;
+  private final boolean isEtagCheckEnabled;
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:89]: nit: old name was not updated

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 21]
```diff
@@ -218,41 +241,50 @@ private void uploadParts() {
...
     stagingFiles.stream()
         // do not upload the file currently being written
-        .filter(f -> closed || !f.equals(currentStagingFile))
+        .filter(f -> closed || !f.file().equals(currentStagingFile))
         // do not upload any files that have already been processed
-        .filter(Predicates.not(multiPartMap::containsKey))
-        .forEach(f -> {
+        .filter(Predicates.not(f -> multiPartMap.containsKey(f.file())))
+        .forEach(fileAndDigest -> {
+          File file = fileAndDigest.file();
```
Comment (jackye1995) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:248]: I think we can name this variable `f` to avoid name changing in subsequent lines

File: aws/src/test/java/org/apache/iceberg/aws/s3/TestS3OutputStream.java [hunk 22]
```diff
@@ -77,14 +86,15 @@
   private final String newTmpDirectory = "/tmp/newStagingDirectory";
 
   private final AwsProperties properties = new AwsProperties(ImmutableMap.of(
-      AwsProperties.S3FILEIO_MULTIPART_SIZE, Integer.toString(5 * 1024 * 1024),
+      AwsProperties.S3FILEIO_MULTIPART_SIZE, Integer.toString(FIVE_MBS),
```
Comment (jackye1995) [inline aws/src/test/java/org/apache/iceberg/aws/s3/TestS3OutputStream.java:89]: this change does not seem necessary

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 16 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:91]: We need both to handle case where written bytes is larger than multipart size but less than multipart threshold. In that case we will have two files, and so two part message digests, whereas we need the whole message digest in put request.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 18 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:200]: This is what I had before, but changed that to instead keep bytes as per @rdblue comment.

**Review by jackye1995 — COMMENTED** [review]
I am back, thanks for the review meanwhile @danielcweeks @rdblue , I share the same comments. Also could you remove all the references of the phrase "etag"? People don't have much context have no idea what that means, and it's better to use "checksum" for those.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 14 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:433]: It needs to be mutable if we are not passing MessageDigest. Earlier I had it mutable as I was passing MessageDigest, which I was able to update while bytes are written. File is created in newStream, but digest bytes are available later when stream is closed.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 21 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:248]: using `f` would lead to `LocalVariableName` checkstyle failure. Previous discussion here https://github.com/apache/iceberg/pull/3813#discussion_r788252447

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 15 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:377]: Makes sense, I can let the callers through the exceptions directly. 

@jackye1995 for logging message, the sensitive part of request like `SSECustomerKey` is already redacted. I also feel just seeing random exception like ProtocolException without much information only makes thing confusing for end users.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 17 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:420]: We can do that for multi part request, but for put request doing this is a bit awkward as either we will have to check for all files or just rely on the existing etag flag. I will update it to use hasDigest never the less for places where we can use it.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 8 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java]: Sure, changed this to RuntimeException.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 18 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:200]: @danielcweeks @rdblue either we can create a new MessageDigest for each file and use MessageDigest in `FileAndDigest` or we can use bytes. Former is likely more readable and later is likely more memory efficient. I feel memory efficiency here is not the biggest concern, so I am happy to go back to what I had originally. However, let me know what you folks think.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 23]
```diff
@@ -218,32 +241,43 @@ private void uploadParts() {
...
           UploadPartRequest uploadRequest = requestBuilder.build();
 
           CompletableFuture<CompletedPart> future = CompletableFuture.supplyAsync(
               () -> {
-                UploadPartResponse response = s3.uploadPart(uploadRequest, RequestBody.fromFile(f));
+                UploadPartResponse response = null;
+                try {
+                  response = s3.uploadPart(uploadRequest, RequestBody.fromFile(file));
+                } catch (UncheckedIOException uncheckedIOException) {
+                  throw new UncheckedIOException(
```
Comment (SinghAsDev) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:270]: I am thinking stripping of message from original `UncheckedIOException` is probably OK as I did not see any helpful info in there, but then it's not guaranteed to remain the same. The security concern expressed by @jackye1995 on logging entire request (which I believe is fine as sensitive parts of requests are already redacted) is also an open conversation.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 16 above)
Comment (rdblue) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:91]: Can we recalculate the digest if we switch to multipart? Or can we at least stop using the complete message digest after switching to multipart?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 21 above)
Comment (rdblue) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:248]: I'd rather suppress the checkstyle rule and have fewer changes.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 23 above)
Comment (rdblue) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:270]: What was the security concern? I don't think that this should be catching the exception and editing it, unless there's a strong reason. Since `UncheckedIOException` is a wrapper produced by other Iceberg code, it's very unlikely to add information that is a security concern.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 14 above)
Comment (rdblue) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:434]: In that case, I think it is a reason to go back to storing the digest. I didn't realize that the digest was getting updated after creating this. Seems like there should be a cleaner way where we update the digest and create an immutable object when the file is closed.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 23 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:270]: Security concern is mostly about s3 request info, previous thread is here https://github.com/apache/iceberg/pull/3813#discussion_r789155286

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 21 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:248]: That's fine with me. Trying to understand the rational though. Is it to avoid merge conflicts for folks that may be doing a cherry-pick?

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 16 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:91]: Yea, I was thinking about this too. We can't do that if we were to use digest stream, I will have to revert to my previous implementation of maintaining and updating digest in writes though.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 21 above)
Comment (rdblue) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:248]: To avoid conflicts.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 23 above)
Comment (rdblue) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:270]: Thanks for the info. I wouldn't worry about it. There is no need for error handling here so let's just let the exception propagate and we can worry about that concern in more appropriate places.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 16 above)
Comment (danielcweeks) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:91]: I don't think we want to try to recalculate if it requires rereading the file in order to calculate the checksum.  I still feel like pushing the per file digest into the FileAndDigest option is the best approach and keep the full file digest for the put object upload as well.  I think we're over optimizing by trying to stop the full file calculation and we could always log the full file MD5.  

There's another option that might simplify things but would require a larger change involving uploads.  If we simply make the first part of the multipart threshold larger (multiplart+threshold) then each FileAndDigest would have a digest associated with it and could be used for the put object or put part uploads.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 18 above)
Comment (danielcweeks) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:200]: I don't think we need to worry too much about the extra bytes necessary to store the digest.  Readability/maintainability is much more important than that little bit of efficiency.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 14 above)
Comment (danielcweeks) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:434]: +1 to Ryan's comment.  Keep a digest per file.  It's a trivial amount of memory and better encapsulated.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java (hunk 16 above)
Comment (SinghAsDev) [reply aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:91]: Actually, just realized we can just turn off the digest updates in `DigestOutputStream` after we have crossed multi part threshold size. Will add that.

File: aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java [hunk 24]
```diff
@@ -209,6 +209,12 @@
    */
   public static final String CLIENT_ASSUME_ROLE_REGION = "client.assume-role.region";
 
+  /**
+   * Enables eTag checks for S3 PUT and MULTIPART upload requests.
+   */
+  public static final String S3_CHECKSUM_ENABLED = "s3.checksum.enabled";
```
Comment (danielcweeks) [inline aws/src/main/java/org/apache/iceberg/aws/AwsProperties.java:215]: I know the conventions for property naming are a little mixed, but I believe this should be `s3.checksum-enabled`

**Comment by danielcweeks** [comment]
LGTM, just waiting on checks

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 25]
```diff
@@ -178,9 +192,27 @@ private void newStream() throws IOException {
...
+
+    stagingFiles.add(new FileAndDigest(currentStagingFile, currentPartMessageDigest));
+
+    if (isChecksumEnabled) {
+      DigestOutputStream digestOutputStream = new DigestOutputStream(new DigestOutputStream(new BufferedOutputStream(
+          new FileOutputStream(currentStagingFile)), currentPartMessageDigest), completeMessageDigest);
 
-    stream = new CountingOutputStream(new BufferedOutputStream(new FileOutputStream(currentStagingFile)));
+      // if switched over to multipart threshold already, no need to update complete message digest
+      if (multipartUploadId != null) {
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:208]: Isn't this known before we create the double `DigestOutputStream`? I think you could just check this above and create a single-wrapped stream instead.

File: aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java [hunk 26]
```diff
@@ -110,6 +118,12 @@
     multiPartSize = awsProperties.s3FileIoMultiPartSize();
     multiPartThresholdSize =  (int) (multiPartSize * awsProperties.s3FileIOMultipartThresholdFactor());
     stagingDirectory = new File(awsProperties.s3fileIoStagingDirectory());
+    isChecksumEnabled = awsProperties.isS3ChecksumEnabled();
+    try {
+      completeMessageDigest = isChecksumEnabled ? MessageDigest.getInstance(digestAlgorithm) : null;
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException("Failed to create message digest needed for s3 checksum checks.", e);
```
Comment (rdblue) [inline aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java:125]: Nit: no need for punctuation at the end of error messages.

**Review by rdblue — APPROVED** [review]

**Comment by danielcweeks** [comment]
Thanks @SinghAsDev !!
